File: ruby/easy/add_binary.rb
# @param {String} a
# @param {String} b
# @return {String}
def add_binary(a, b)
    i, j, carry = a.length-1, b.length-1, 0
    obj = { sum: 0, carry: 0, str: "" }
    while i >= 0 and j >= 0
        add_to_str(obj, (a[i].ord - '0'.ord) + (b[j].ord - '0'.ord))
        i -= 1
        j -= 1
    end

    while i >= 0
        add_to_str(obj, a[i].ord - '0'.ord)
        i -= 1
    end

    while j >= 0
        add_to_str(obj, b[j].ord - '0'.ord)
        j -= 1
    end

    add_to_str(obj, 0) if obj[:carry] > 0
    return obj[:str]
end

def add_to_str(obj, sum)
    obj[:sum] = sum + obj[:carry]
    obj[:carry] = (obj[:sum] > 1) ? 1 : 0
    obj[:str] = (((obj[:sum] % 2) != 0) ? "1" : "0") + obj[:str];
end

File: javascript/easy/addBinary.js
/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
var addBinary = function(a, b) {
    let i = a.length-1, j = b.length-1, obj = { sum: 0, carry: 0, str: "" };
    while(i >= 0 && j >= 0) {
        addToStr(obj, (a[i] - '0') + (b[j] - '0'));
        i--; j--;
    }
    while(i >= 0) {
        addToStr(obj, (a[i] - '0'));
        i--;
    }
    while(j >= 0) {
        addToStr(obj, (b[j] - '0'));
        j--;
    }
    if(obj.carry > 0) {
        addToStr(obj, 0);
    }
    return obj.str;
};

var addToStr = function(obj, val) {
    let sum = (val + obj.carry);
    obj.carry = (sum > 1) ? 1 : 0;
    obj.str = ((sum % 2) != 0 ? "1" : "0") + obj.str;
};
